I3uckwheat/cards-of-carousal
# Bug/282 player join during selection phase

#### 0. Make sure your issue is linked:

Close #282 

#### 1. Purpose:


Currently if a player joins mid-game while other players are selecting their cards, the game can get stuck waiting for the new player to submit cards they don't have. This fix will deal new cards to the new player and allow them to play right away, thus keeping the game moving.

#### 2. Implementation:


This one required a little bit of a rework in multiple areas:

### HostReducer
* PLAYER_CONNECTED
  * Check if the game state is waiting for players to submit cards
    * If so, deal cards again 
      * current players will still only have the cards they previously had, the new one will receive a full hand
    * If not, proceed as normal
  * Decouples clearing submitted cards from the `dealWhiteCards` function so we can retain which players have submitted cards
  * Imports hostReducerMiddleware to send the cards to the players again once they've been updated. This function is only a side effect and does not fallthrough to a dispatch, so it is a safe import.
    * When sending cards, implements a `newPlayer` property to indicate whether the cards are being dealt as a result of a new player joining. This ensures that the middleware sends the correct socket message to the players (see next).

### hostReducerMiddleware
* `sendCardsToPlayers`
  * When sending cards, implements a `shouldPreserveGameState` property to indicate whether the players receiving cards should be brought to the player-select screen. This boolean (calculated based on the new player's ID compared to the player receiving cards) ensures that players who have already submitted their cards aren't brought back to the selection screen.

### PlayerReducer
Looks to the payload to indicate whether the player's game state should change to 'player-select' (see above)

### HostScreenController
Removes a game state case that should not have been added to the host end. 

#### 3. Possible Issues:


Consider additional edge cases related to the game state and new players joining. Try joining with multiple players mid game.

#### 4. How to Test:


1. Pull branch
2. Run game, open 5+ tabs
3. Join with 3 players, start game, submit cards on one of the tabs
4. Join game with an additional player (or more).
5. Observe the game state of each player (and host), attempt to continue on with the game.

#### 5. Cleanup
